File: cses/07_mathematics/1729_stick-game/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    k = int(data[1])
    moves = tuple(int(data[2 + i]) for i in range(k))
    maxp = max(moves)

    # win[i] == 1 -> thế THẮNG; 0 -> thế THUA (giá trị mặc định).
    # Nới rộng thêm maxp để chỉ số win[i + p] luôn hợp lệ, khỏi cần kiểm tra biên.
    win = bytearray(n + maxp + 1)
    res = bytearray(n)  # ký tự kết quả cho các vị trí 1..n

    W = 87  # ord('W')
    L = 76  # ord('L')

    mv = moves
    w = win
    r = res
    # Lan truyền tiến: duyệt i tăng dần nên khi đọc win[i] mọi thế thua j < i đã
    # đánh dấu xong các thế thắng của nó.
    for i in range(n + 1):
        if w[i]:
            # Thế THẮNG (i >= 1 vì win[0] mặc định 0 là thế thua).
            r[i - 1] = W
        else:
            # Thế THUA: mọi trạng thái i + p (p thuộc P) đều là thế THẮNG.
            for p in mv:
                w[i + p] = 1
            if i:
                r[i - 1] = L

    sys.stdout.write(r.decode())


main()

File: cses/07_mathematics/1729_stick-game/solution.cpp
#include <algorithm>
#include <cstddef>
#include <iostream>
#include <string>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::size_t n, k;
    std::cin >> n >> k;
    std::vector<std::size_t> moves(k);
    for (std::size_t& move : moves) {
        std::cin >> move;
    }

    // Nới rộng thêm largest_move phần tử để khi đánh dấu winning[sticks + move]
    // luôn hợp lệ, khỏi cần kiểm tra biên. Mặc định 0 nghĩa là thế THUA.
    const std::size_t largest_move =
        *std::max_element(moves.begin(), moves.end());
    std::vector<unsigned char> winning(n + largest_move + 1, 0);
    std::string answer(n, 'L');

    // Lan truyền tiến: duyệt tăng dần nên khi tới sticks mọi thế thua j < sticks
    // đã đánh dấu xong các thế thắng của nó.
    for (std::size_t sticks = 0; sticks <= n; ++sticks) {
        if (winning[sticks] != 0) {
            // Thế THẮNG: đã có nước đi đưa đối thủ về một thế thua trước đó.
            if (sticks > 0) {
                answer[sticks - 1] = 'W';
            }
            continue;
        }
        // Thế THUA: mọi trạng thái bốc được để quay về sticks đều là thế THẮNG.
        for (const std::size_t move : moves) {
            winning[sticks + move] = 1;
        }
    }

    std::cout << answer << '\n';
    return 0;
}
